text/0000-system-buildpacks-in-builder.md

```diff
@@ -0,0 +1,134 @@
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a mechanism that would allow a builder to contain a default set of buildpacks that run on every build, independent of which buildpack group passes detection.
+
+# Definitions
+[definitions]: #definitions
+
+* _system buildpacks_ - a standard buildpack, conforming to the Buildpack API, but which runs outside of normal groups
+
+# Motivation
+[motivation]: #motivation
+
+Forthcoming changes to the lifecycle (such as [removal of shell-specific logic](https://github.com/buildpacks/rfcs/pull/168)) will remove capabilities that users have come to expect. This includes mechanisms like `.profile`, which allows a buildpack user to customize the environment a process type runs in. We seek to replace these lost mechanisms with buildpacks, in an effort to preserve the capability while still removing complexity from the lifecycle.
+
+# What it is
+[what-it-is]: #what-it-is
+
+We introduce a `[system]` table in the `builder.toml` schema with the following structure:
+
+```
+[[system.pre.buildpacks]]
+  id = "<buildpack ID>"
+  version = "<buildpack version>"
+  optional = true
+
+[[system.post.buildpacks]]
+  id = "<buildpack ID>"
+  version = "<buildpack version>"
+  optional = true
+```
+
+The fields in the `system.pre.buildpacks` table and `system.post.buildpacks` table match the fields in the existing [`order.group` table](https://buildpacks.io/docs/reference/config/builder-config/#order-_list-required_). However, `optional` in this case is required and the only acceptable value is `true`. Non-optional buildpacks will cause the builder creation to fail.
+
+When a builder includes one or more `system.*.buildpacks` entry, each build phase that uses the builder will run the `pre` buildpacks before ALL other buildpacks (whether they are defined by a default `order.group` or if they are provided as an explicit order to the lifecycle) and the `post` buildpack after ALL other buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Unless otherwise stated, system buildpacks conform to the [buildpack API](https://github.com/buildpacks/spec/blob/main/buildpack.md).
+
+The `system.pre.buildpacks` will be transformed into a new table in `order.toml`. The `[system]` table in `order.toml` will be processed by the lifecycle, and each `pre`/`post` buildpack will run during the detect phase. Those that pass detection will run during the build phase.
+
+## Detection
+
+The exit code of detection by system buildpacks MUST NOT influence the selected buildpack group. If no system buildpacks pass detection, any buildpack group MAY pass detection. If a system buildpack passes detection and no buildpack groups pass detection, then detection MUST fail.
+
+System buildpacks may require/provide in the build plan following standard buildpack API specification.
+
+A new flag to the lifecycle `detector`, `--disable-system-buildpacks`, will disable system buildpacks.
```

Comment:
Instead of having a `--disable-system-buildpacks` flag, once system buildpacks are in a separate file could an empty `CNB_SYSTEM_PATH=` or `-system=""`or something like that indicate the same thing without requiring another flag. Therefore we wouldn't run into weird situations where the platform is both explicitly passing the path to a system buildpack file and also disabling those same buildpacks.

Reply:
addressed in f8f776d